<!DOCTYPE HTML>
<html lang="en" class="light" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Find highly modified motif sequences - Modkit</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="custom.css">

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    </head>
    <body class="sidebar-visible no-js">
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('light')
            html.classList.add(theme);
            var body = document.querySelector('body');
            body.classList.remove('no-js')
            body.classList.add('js');
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var body = document.querySelector('body');
            var sidebar = null;
            var sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            body.classList.remove('sidebar-visible');
            body.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="quick_start.html"><strong aria-hidden="true">1.</strong> Quick Start guides</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="intro_pileup.html"><strong aria-hidden="true">1.1.</strong> Constructing bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_pileup_hemi.html"><strong aria-hidden="true">1.2.</strong> Make hemi-methylation bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_adjust.html"><strong aria-hidden="true">1.3.</strong> Updating and adjusting MM tags</a></li><li class="chapter-item expanded "><a href="intro_sample_probs.html"><strong aria-hidden="true">1.4.</strong> Inspecting base modification probabilities</a></li><li class="chapter-item expanded "><a href="intro_summary.html"><strong aria-hidden="true">1.5.</strong> Summarizing a modBAM</a></li><li class="chapter-item expanded "><a href="intro_stats.html"><strong aria-hidden="true">1.6.</strong> Calculating modification statistics in regions</a></li><li class="chapter-item expanded "><a href="intro_call_mods.html"><strong aria-hidden="true">1.7.</strong> Calling mods in a modBAM</a></li><li class="chapter-item expanded "><a href="intro_edge_filter.html"><strong aria-hidden="true">1.8.</strong> Removing modification calls at the ends of reads</a></li><li class="chapter-item expanded "><a href="intro_repair.html"><strong aria-hidden="true">1.9.</strong> Repair MM/ML tags on trimmed reads</a></li><li class="chapter-item expanded "><a href="intro_motif.html"><strong aria-hidden="true">1.10.</strong> Working with sequence motifs</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="intro_motif_bed.html"><strong aria-hidden="true">1.10.1.</strong> Making a motif BED file</a></li><li class="chapter-item expanded "><a href="intro_find_motifs.html" class="active"><strong aria-hidden="true">1.10.2.</strong> Find highly modified motif sequences</a></li><li class="chapter-item expanded "><a href="evaluate_motif.html"><strong aria-hidden="true">1.10.3.</strong> Evaluate and refine a table of known motifs</a></li><li class="chapter-item expanded "><a href="motif_search_structured_logging.html"><strong aria-hidden="true">1.10.4.</strong> Structured logging schema</a></li></ol></li><li class="chapter-item expanded "><a href="intro_extract.html"><strong aria-hidden="true">1.11.</strong> Extracting read information to a table</a></li><li class="chapter-item expanded "><a href="intro_localize.html"><strong aria-hidden="true">1.12.</strong> Investigating patterns with localise</a></li><li class="chapter-item expanded "><a href="intro_dmr.html"><strong aria-hidden="true">1.13.</strong> Perform differential methylation scoring</a></li><li class="chapter-item expanded "><a href="intro_validate.html"><strong aria-hidden="true">1.14.</strong> Validate ground truth results</a></li><li class="chapter-item expanded "><a href="intro_entropy.html"><strong aria-hidden="true">1.15.</strong> Calculating methylation entropy</a></li><li class="chapter-item expanded "><a href="intro_include_bed.html"><strong aria-hidden="true">1.16.</strong> Narrow output to specific positions</a></li><li class="chapter-item expanded "><a href="intro_bedmethyl_merge.html"><strong aria-hidden="true">1.17.</strong> Manipulate bedMethyl files</a></li><li class="chapter-item expanded "><a href="intro_modbam_check_tags.html"><strong aria-hidden="true">1.18.</strong> Check modified base tags</a></li><li class="chapter-item expanded "><a href="intro_open_chromatin.html"><strong aria-hidden="true">1.19.</strong> Find regions of accessible chromatin</a></li></ol></li><li class="chapter-item expanded "><a href="advanced_usage.html"><strong aria-hidden="true">2.</strong> Extended subcommand help</a></li><li class="chapter-item expanded "><a href="troubleshooting.html"><strong aria-hidden="true">3.</strong> Troubleshooting</a></li><li class="chapter-item expanded "><a href="faq.html"><strong aria-hidden="true">4.</strong> Frequently asked questions</a></li><li class="chapter-item expanded "><a href="limitations.html"><strong aria-hidden="true">5.</strong> Current limitations</a></li><li class="chapter-item expanded "><a href="perf_considerations.html"><strong aria-hidden="true">6.</strong> Performance considerations</a></li><li class="chapter-item expanded "><a href="algo_details.html"><strong aria-hidden="true">7.</strong> Algorithm details</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering.html"><strong aria-hidden="true">7.1.</strong> Pass/fail base modification calls</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering_details.html"><strong aria-hidden="true">7.1.1.</strong> Threshold examples</a></li><li class="chapter-item expanded "><a href="filtering_numeric_details.html"><strong aria-hidden="true">7.1.2.</strong> Numeric details</a></li></ol></li><li class="chapter-item expanded "><a href="dmr_scoring_details.html"><strong aria-hidden="true">7.2.</strong> DMR model and scoring details</a></li><li class="chapter-item expanded "><a href="collapse.html"><strong aria-hidden="true">7.3.</strong> Ignoring and combining calls</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <!-- Track and set sidebar scroll position -->
        <script>
            var sidebarScrollbox = document.querySelector('#sidebar .sidebar-scrollbox');
            sidebarScrollbox.addEventListener('click', function(e) {
                if (e.target.tagName === 'A') {
                    sessionStorage.setItem('sidebar-scroll', sidebarScrollbox.scrollTop);
                }
            }, { passive: true });
            var sidebarScrollTop = sessionStorage.getItem('sidebar-scroll');
            sessionStorage.removeItem('sidebar-scroll');
            if (sidebarScrollTop) {
                // preserve sidebar scroll position when navigating via links within sidebar
                sidebarScrollbox.scrollTop = sidebarScrollTop;
            } else {
                // scroll sidebar to current active section when navigating via "next/previous chapter" buttons
                var activeSection = document.querySelector('#sidebar .active');
                if (activeSection) {
                    activeSection.scrollIntoView({ block: 'center' });
                }
            }
        </script>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Modkit</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="find-highly-modified-motif-sequences"><a class="header" href="#find-highly-modified-motif-sequences">Find highly modified motif sequences</a></h1>
<p>The <code>modkit find-motifs</code> command will attempt to summarize short genome sequences (motifs) that are more found to be highly modified (i.e. enriched for methylation).
The input to this command is a bedMethyl generated by <code>modkit pileup</code> and the reference sequence used.
For example, to run the command with default settings (recommended):</p>
<pre><code class="language-bash">bedmethyl=/path/to/pileup.bed
ref=/path/to/reference.fasta

modkit motif search \
  -i ${bedmethyl} \
  -r ${ref} \
  -o ./motifs.tsv \
  --threads 32 \
  --log ./modkit_find_motifs_log.txt 
</code></pre>
<p>Specifying an output with <code>-o</code> will generate a machine-readable tab-separated-values file, a human-readable version of the table will always be logged to the terminal and the logfile.</p>
<p>If you have compressed your bedMethyl table with <code>bgzip</code> you can use the compressed table as well.
A compressed table with <code>tabix</code>-generated index is required to use the <code>--contig</code> option which will find motifs using only a single contig in the bedMethyl table.
This may be useful for applications such as metagenomics.</p>
<h2 id="output-format"><a class="header" href="#output-format">Output format</a></h2>
<p>All output tables are output in two formats, machine-readable and human-readable.
The human-readable tables are always output to the log and terminal, the machine-readable tables are output to files specified on the command line.</p>
<h3 id="machine-readable-table"><a class="header" href="#machine-readable-table">Machine-readable table</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>mod_code</td><td>code specifying the modification found in the motif</td><td>str</td></tr>
<tr><td>2</td><td>motif</td><td>sequence of identified motif using <a href="https://www.bioinformatics.org/sms/iupac.html">IUPAC</a> codes</td><td>str</td></tr>
<tr><td>3</td><td>offset</td><td>0-based offset into the motif sequence of the modified base</td><td>int</td></tr>
<tr><td>4</td><td>frac_mod</td><td>fraction of time this sequence is found in the <em>high modified</em> set col-5 / (col-5 + col-6)</td><td>float</td></tr>
<tr><td>5</td><td>high_count</td><td>number of occurances of this sequence in the <em>high-modified</em> set</td><td>int</td></tr>
<tr><td>6</td><td>low_count</td><td>number of occurances of this sequence in the <em>low-modified</em> set</td><td>int</td></tr>
<tr><td>7</td><td>mid_count</td><td>number of occurances of this sequence in the <em>mid-modified</em> set</td><td>int</td></tr>
</tbody></table>
</div>
<h3 id="human-readable-table"><a class="header" href="#human-readable-table">Human-readable table</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>motif</td><td>human-readable representation of the motif sequence with the modification code in brackets</td><td>str</td></tr>
<tr><td>2</td><td>frac_mod</td><td>fraction of time this sequence is found in the <em>high modified</em> set col-3 / (col-3 + col-4)</td><td>float</td></tr>
<tr><td>3</td><td>high_count</td><td>number of occurances of this sequence in the <em>high-modified</em> set</td><td>int</td></tr>
<tr><td>4</td><td>low_count</td><td>number of occurances of this sequence in the <em>low-modified</em> set</td><td>int</td></tr>
<tr><td>5</td><td>mid_count</td><td>number of occurances of this sequence in the <em>mid-modified</em> set</td><td>int</td></tr>
</tbody></table>
</div>
<h2 id="specifying-known-motifs"><a class="header" href="#specifying-known-motifs">Specifying known motifs</a></h2>
<p>Multiple motif sequences suspected to be present can be specified with the <code>--known-motif</code> option.
A machine-readable table of the motif sequences that are not found during the search can be specified with the <code>--known-motifs-table</code> option.
Using this option will add two columns to the above tables:</p>
<div class="table-wrapper"><table><thead><tr><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>status</td><td>equal, Subset, Superset, or Disjoint describes the relationship of the discovered motif to the known motif</td><td>str</td></tr>
<tr><td>closest_known_motif</td><td>of all motifs specified with <code>--known-motif</code> the one that is most similar to the discovered motif</td><td>str</td></tr>
</tbody></table>
</div>
<p>If any of the known motifs are not found during the search process an additional table is also emitted in machine- and human-readable versions.</p>
<h3 id="machine-readable-table-1"><a class="header" href="#machine-readable-table-1">Machine-readable table</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>mod_code</td><td>code specifying the modification found in the motif</td><td>str</td></tr>
<tr><td>2</td><td>motif</td><td>sequence of identified motif using <a href="https://www.bioinformatics.org/sms/iupac.html">IUPAC</a> codes</td><td>str</td></tr>
<tr><td>3</td><td>offset</td><td>0-based offset into the motif sequence of the modified base</td><td>int</td></tr>
<tr><td>4</td><td>frac_mod</td><td>fraction of time this sequence is found in the <em>high modified</em> set col-5 / (col-5 + col-6)</td><td>float</td></tr>
<tr><td>5</td><td>high_count</td><td>number of occurances of this sequence in the <em>high-modified</em> set</td><td>int</td></tr>
<tr><td>6</td><td>low_count</td><td>number of occurances of this sequence in the <em>low-modified</em> set</td><td>int</td></tr>
<tr><td>7</td><td>mid_count</td><td>number of occurances of this sequence in the <em>mid-modified</em> set</td><td>int</td></tr>
<tr><td>8</td><td>status</td><td>equal, Subset, Superset, or Disjoint describes the relationship of the known motif to the closest discovered motif</td><td>str</td></tr>
<tr><td>9</td><td>closest_found_motif</td><td>which of the discovered motifs is most simuilar to the known motif</td><td>str</td></tr>
</tbody></table>
</div>
<h3 id="human-readable-table-1"><a class="header" href="#human-readable-table-1">Human-readable table</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>motif</td><td>human-readable representation of the motif sequence with the modification code in brackets</td><td>str</td></tr>
<tr><td>2</td><td>frac_mod</td><td>fraction of time this sequence is found in the <em>high modified</em> set col-3 / (col-3 + col-4)</td><td>float</td></tr>
<tr><td>3</td><td>high_count</td><td>number of occurances of this sequence in the <em>high-modified</em> set</td><td>int</td></tr>
<tr><td>4</td><td>low_count</td><td>number of occurances of this sequence in the <em>low-modified</em> set</td><td>int</td></tr>
<tr><td>5</td><td>mid_count</td><td>number of occurances of this sequence in the <em>mid-modified</em> set</td><td>int</td></tr>
<tr><td>6</td><td>status</td><td>equal, Subset, Superset, or Disjoint describes the relationship of the known motif to the closest discovered motif</td><td>str</td></tr>
<tr><td>7</td><td>closest_found_motif</td><td>which of the discovered motifs is most simuilar to the known motif</td><td>str</td></tr>
</tbody></table>
</div>
<h2 id="simple-description-of-the-search-algorithm"><a class="header" href="#simple-description-of-the-search-algorithm">Simple description of the search algorithm</a></h2>
<p>The first step in <code>find-motifs</code> is to categorize each genomic position in the pileup into one of three groups based on the <code>fraction modified</code> <a href="./intro_pileup.html#bedmethyl-column-descriptions">column in the bedMethyl</a>:</p>
<ul>
<li>Low-modified</li>
<li>Mid-modified</li>
<li>High-modified</li>
</ul>
<p>The threshold values for each group can be set on the command line (<code>--high-thresh</code> and <code>--low-thresh</code>).
For example, consider a high threshold of 0.6 and a low threshold of 0.2 the following 3 bedMethyl records would be put into the high-, low-, and mid-groups, respectively:</p>
<pre><code class="language-text">contig1      6       7       a       27      -       6       7       255,0,0 27      96.30   26      1       0       0       3       0       0
contig1      8       9       a       24      -       8       9       255,0,0 24      4.17    1       23      0       0       5       1       0
contig1      218     219     a       21      +       218     219     255,0,0 21      28.57   6       15      0       2       3       0       0
</code></pre>
<p>The sequence around each modified position is then collected from the reference FASTA file.
The length of the sequence can be set with the <code>--context-size</code> option, accepting two values: the number of bases upstream and the number of bases downstream of the modified location.
For example <code>--context-size 12 12</code> will collect 12 bases upstream and 12 bases downstream of the modified base for a maximum motif length of 25 base pairs.
The algorithm then iteratively expands, contracts, and merges sequences while the following criteria are met:</p>
<ul>
<li>The number of occurrences in the high-modified set is greater than <code>min_sites</code> (set by <code>--min-sites</code>).</li>
<li>The fraction \( \frac{\textit{H}}{\textit{H} + \textit{L}} \), is greater than <code>frac_mod</code> (set with <code>--min-frac-mod</code>),  where \( \textit{H} \) and \( \textit{L} \) is the number of total sequence contexts in the high-modified and low-modified set, respectively.</li>
<li>The log-odds of the context being in the high-modified category is greater than <code>min_log_odds</code> (set with <code>--min-log-odds</code>).</li>
</ul>
<p>Once a motif sequence cannot be changed (made more general or more restrictive) without violating one of these criteria, the motif sequence is considered complete.
As the algorithm continues, context sequences that match discovered sequences are removed from consideration.</p>
<p>A secondary search step, called the "exhaustive search", is also performed by starting with every k-mer (where k is less than the total sequence length, 3 by default, set with <code>--exhaustive-seed-len</code>) at every motif position.
The log-odds threshold for this search is usually higher and set with <code>--exhaustive-seed-min-log-odds</code>.
Decreasing this value can drastically increase computational time, see the next section for more details.</p>
<h2 id="options-to-decrease-or-limit-search-time"><a class="header" href="#options-to-decrease-or-limit-search-time">Options to decrease or limit search time.</a></h2>
<p>The optional (but recommended) exhaustive search step checks many sparsely defined sequences, called seeds, for enrichment as defined by <code>--exhaustive-seed-min-log-odds</code>.
The algorithm then refines the seeds above the threshold into candidate motifs using the method described <a href="intro_find_motifs.html#simple-description-of-the-search-algorithm">previously</a>.
An example seed is <code>GNNNNNNNNNNC[m]NNNNNNNNNNNA</code> using the modification code for 5mC.
The number of seeds is defined by the <code>--context-size</code> and <code>--exhaustive-seed-length</code> parameters.
For example, if <code>--context-size a b</code> is passed the seeds will be length \( a + b + 1 \).
The default value is <code>a = 12</code>, <code>b = 12</code> so the sequence length is 25, but we only use 24 of those positions, call this \( n \).
The default value for <code>--exhaustive-seed-length</code> is 3, call this \( k \).
Remember that there are 4 primary sequence bases.
So the number of seeds to search is \( k^4 * \binom{n}{k} \) or 129,536 seeds.
Most of these seeds will fall below the log-odds required, and not follow-on to refinement.
However, if a large number ( \( \gt 500 \) ) are above the log-odds threshold then this step can be very time consuming.
There are a couple options that can decrease or limit this run time, whist usually affording identical or very good results.</p>
<ol>
<li>A simple timeout.
When the search starts, the top-N (as ranked by log-odds) seeds are taken.
The search is run on this "batch", if the timeout has not expired the next batch of N is taken.
If all of the seeds are not evaluated before running out of time, the results are returned but an ERROR message is displayed.</li>
</ol>
<pre><code class="language-bash"># specify that search should stop after 15 minutes
$ modkit motif search ... --search-timeout 15m
</code></pre>
<ol start="2">
<li>Top percent.
Only search the top P-% of seeds (or a maximum number).
So a command like this <code>--search-top-pct 20 --max-exhaustive-seeds 200</code> would search the top 20% of seeds as ranked by their log-odds for being motifs or the top 200 seeds if 20% of all of the seeds passing <code>--exhaustive-seed-min-log-odds</code> is greater than 200.
<strong>n.b.</strong> With default settings, in the most pathological settings 129,536 could be searched.</li>
</ol>
<pre><code class="language-bash"># only search the min(top 20% of seeds, 100 seeds)
$ modkit motif search ... --search-top-pct 20 --max-exhaustive-seeds 100
</code></pre>
<ol start="3">
<li>Batch and Narrow Optimization.
This algorithm is a combination of (1) and (2).
In this scenario, we take the top P-% of seeds (or the maximum allowed) as in (2).
But instead of stopping as in (2) we continue (as in (1)) but we remove contexts from consideration based on the motifs we've found before the next round.
At the end of the round, if zero motifs are found, we stop.</li>
</ol>
<pre><code class="language-bash"># same as (2), but continue to search after each batch
$ modkit motif search .. --search-top-pct 20 --max-exhaustive-seeds 100 --narrow-search
</code></pre>
<ol start="4">
<li>Batch and Narrow Optimization with Timeout.
This scenario is the same as (3), but if we've run out of time at the end of a round - stop.</li>
</ol>
<pre><code class="language-bash"># same as (3), but continue to search after each batch, until we've spent 15 minutes searching.
$ modkit motif search .. --search-top-pct 20 --max-exhaustive-seeds 100 --narrow-search --search-timeout 15m
</code></pre>
<h2 id="tuning-parameters-and---skip-search"><a class="header" href="#tuning-parameters-and---skip-search">Tuning parameters and <code>--skip-search</code></a></h2>
<p>The default parameters have been picked to be sufficiently sensitive, however if you decide to adjust the parameters in general increasing sensitivity will increase compute time.</p>
<ol>
<li>Increasing the <code>--min-frac-mod</code> will stop search earlier which will decrease compute time.</li>
<li>Decreasing <code>--min-sites</code> has the largest effect and can especially cause the secondary search to crawl more sequences.
Decreasing <code>--min-sites</code> along with <code>--skip-search</code> may be a useful technique to find very rare sequence motifs.</li>
<li>Increasing <code>--exhaustive-seed-min-log-odds</code> can drastically decrease compute time (sometimes while maintaining sensitivity).</li>
</ol>
<p>Also consider the additional steps in <a href="./perf_considerations.html#parallelism-in-find-motifs-when-to---skip-search">performance considerations.</a></p>
<h2 id="investigating-how-motifs-are-evaluated"><a class="header" href="#investigating-how-motifs-are-evaluated">Investigating how motifs are evaluated</a></h2>
<p>The <code>motif search</code> function uses JSON-lines structured logging that can be searched and queried for specific events.
The schema and some examples are described <a href="./motif_search_structured_logging.html">in another section</a>.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="intro_motif_bed.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="evaluate_motif.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="intro_motif_bed.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="evaluate_motif.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
